src/AdvancedMH.jl

```diff
@@ -0,0 +1,135 @@
+module AdvancedMH
+
+# Import the relevant libraries.
+using Reexport
+using Turing.Interface
+@reexport using Distributions
+using Random
+
+# Import specific functions and types to use or overload.
+import Distributions: VariateForm, ValueSupport, variate_form, value_support, Sampleable
+import MCMCChains: Chains
+import Turing.Interface: step!, AbstractSampler, AbstractTransition, transition_type
+
+# Exports
+export MetropolisHastings, DensityModel, sample
+
+# Define a sampler type.
+struct MetropolisHastings{T} <: AbstractSampler 
+    init_θ :: T
+    proposal :: Function
```

Comment:
Why does this have to be a Function? I find this confusing, especially as you always access the proposal distribution the function returns. Can we change this to a Distribution? Or is the idea that this could be a NN, e.g. normalising flow (but that's in case of Bijectors again a Distributions).

Reply:
I agree that restricting this to be a `Function` is a bit odd, and agree that requiring it to be a distribution of some kind might be more appropriate. Alternatively, we could not restrict at all for now, until we're sure what we actually want. e.g.
```julia
struct MetropolisHastings{Tinit, Tproposal} <: AbstractSampler
    init_θ::Tinit
    proposal::Tproposal
end
```

Reply:
I'm a bit confused here. From the example, what should be passed in is `x -> MvNormal(x, 0.5)`. It's like a distribution construcotor, or proposal constructor? Instead of an actual proposal distribution. Is my understanding correct?

Reply:
I'm not sure what the alternative is here -- how would you dynamically construct the conditional proposal distribution if the proposal wasn't a function? In this case I'm aiming for RWMH, which requires a new proposal distribution for each parameter draw. There is not (to my knowledge) an easy way to pass a general distribution that can be reparameterized when taking conditionals. Functions make this easy and fast. 

The only alternative I see still requires the generation of an anonymous function, and it would be significantly less general than the functional requirement.

Reply:
Just to be clear.
- I agree it should be a function.
    - We need to parametrise the struct with function as Will suggests though.
- The name `proposal` might not be accurate? I would feel what being constructed is the proposal, but not the function itself.

Reply:
Maybe consider separating the proposal into 

- `θ = θ_prev + Δθ`

where `Δθ` is sampled from the `proposal` distribution.  The `propose` function needs to be modified so that it returns `θ + Δθ` as output. This way we can fix the `proposal`, with potential efficiency improvements.